[BUG] NoClassDefFoundError with caller classloader off in GpuShuffleCoalesceIterator in local-cluster

**Describe the bug**
running `test_cartesian_join_special_case_count` fails with 
`Caused by: java.lang.NoClassDefFoundError: com/nvidia/spark/rapids/Arm`
<details><pre>
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.$anonfun$next$3(GpuShuffleCoalesceExec.scala:214)
E                       at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                       at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.withResource(GpuShuffleCoalesceExec.scala:191)
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.$anonfun$next$2(GpuShuffleCoalesceExec.scala:213)
E                       at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                       at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.withResource(GpuShuffleCoalesceExec.scala:191)
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.$anonfun$next$1(GpuShuffleCoalesceExec.scala:207)
E                       at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                       at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.withResource(GpuShuffleCoalesceExec.scala:191)
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.next(GpuShuffleCoalesceExec.scala:206)
E                       at com.nvidia.spark.rapids.GpuShuffleCoalesceIterator.next(GpuShuffleCoalesceExec.scala:191)
E                       at com.nvidia.spark.rapids.GpuHashAggregateIterator.aggregateInputBatches(aggregate.scala:283)
E                       at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$next$2(aggregate.scala:238)
E                       at scala.Option.getOrElse(Option.scala:189)
E                       at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:235)
E                       at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:181)
E                       at com.nvidia.spark.rapids.ColumnarToRowIterator.$anonfun$fetchNextBatch$2(GpuColumnarToRowExec.scala:241)
E                       at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
E                       at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
E                       at com.nvidia.spark.rapids.ColumnarToRowIterator.withResource(GpuColumnarToRowExec.scala:187)
E                       at com.nvidia.spark.rapids.ColumnarToRowIterator.fetchNextBatch(GpuColumnarToRowExec.scala:238)
E                       at com.nvidia.spark.rapids.ColumnarToRowIterator.loadNextBatch(GpuColumnarToRowExec.scala:215)
E                       at com.nvidia.spark.rapids.ColumnarToRowIterator.hasNext(GpuColumnarToRowExec.scala:255)
E                       at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
E                       at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:349)
E                       at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
E                       at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
E                       at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
E                       at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
E                       at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
E                       at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
E                       at org.apache.spark.scheduler.Task.run(Task.scala:131)
E                       at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
E                       at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)
E                       at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
E                       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E                       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E                       ... 1 more
</pre></details>
 

**Steps/Code to reproduce bug**
invoke:
```bash
TEST_PARALLEL=0 \
  PYSP_TEST_spark_rapids_force_caller_classloader=false \
  NUM_LOCAL_EXECS=1 \
  ./integration_tests/run_pyspark_from_build.sh -k test_cartesian_join_special_case_count
```

**Expected behavior**
Should pass

**Environment details (please complete the following information)**
 - Environment location: Standalone, 
 
 - **Additional context**

Add any other context about the problem here.

originally reported by h/t @pxLi

--- a/jenkins/spark-premerge-build.sh
+++ b/jenkins/spark-premerge-build.sh
@@ -73,6 +73,14 @@ mvn_verify() {
 
     # Triggering here until we change the jenkins file
     rapids_shuffle_smoke_test
+
+    # non-caller classloader smoke test in pseudo-distributed
+    # standalone cluster
+    echo "Running test_cartesian_join_special_case_count with spark.rapids.force.caller.classloader=false"
+    PYSP_TEST_spark_rapids_force_caller_classloader=false \
+        NUM_LOCAL_EXECS=1 \
+        TEST_PARALLEL=0 \
+        ./integration_tests/run_pyspark_from_build.sh -k 'test_cartesian_join_special_case_count[100]'
 }
 
 rapids_shuffle_smoke_test() {

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala
@@ -18,7 +18,7 @@ package com.nvidia.spark.rapids
 
 import java.util
 
-import ai.rapids.cudf.{HostConcatResultUtil, HostMemoryBuffer, JCudfSerialization, NvtxColor, NvtxRange}
+import ai.rapids.cudf.{HostMemoryBuffer, JCudfSerialization, NvtxColor, NvtxRange}
 import ai.rapids.cudf.JCudfSerialization.{HostConcatResult, SerializedTableHeader}
 import com.nvidia.spark.rapids.shims.ShimUnaryExecNode
 
@@ -103,7 +103,7 @@ class HostShuffleCoalesceIterator(
       val firstHeader = serializedTables.peekFirst().header
       if (firstHeader.getNumColumns == 0) {
         (0 until numTablesInBatch).foreach(_ => serializedTables.removeFirst())
-        HostConcatResultUtil.rowsOnlyHostConcatResult(numRowsInBatch)
+        cudf_utils.HostConcatResultUtil.rowsOnlyHostConcatResult(numRowsInBatch)
       } else {
         val headers = new Array[SerializedTableHeader](numTablesInBatch)
         withResource(new Array[HostMemoryBuffer](numTablesInBatch)) { buffers =>
@@ -211,7 +211,7 @@ class GpuShuffleCoalesceIterator(iter: Iterator[HostConcatResult],
         // generate GPU data from batches that are empty.
         GpuSemaphore.acquireIfNecessary(TaskContext.get(), semWaitTime)
         withResource(new MetricRange(opTimeMetric)) { _ =>
-          val batch = HostConcatResultUtil.getColumnarBatch(hostConcatResult, dataTypes)
+          val batch = cudf_utils.HostConcatResultUtil.getColumnarBatch(hostConcatResult, dataTypes)
           outputBatchesMetric += 1
           outputRowsMetric += batch.numRows()
           batch

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala
@@ -16,7 +16,7 @@
 
 package com.nvidia.spark.rapids
 
-import ai.rapids.cudf.{HostConcatResultUtil, NvtxColor, NvtxRange}
+import ai.rapids.cudf.{NvtxColor, NvtxRange}
 import ai.rapids.cudf.JCudfSerialization.HostConcatResult
 import com.nvidia.spark.rapids.shims.{GpuHashPartitioning, ShimBinaryExecNode}
 
@@ -345,7 +345,7 @@ object GpuShuffledHashJoinExec extends Arm {
     // we can bring the build batch to the GPU now
     withResource(hostConcatResult) { _ =>
       buildTime.ns {
-        HostConcatResultUtil.getColumnarBatch(hostConcatResult, dataTypes)
+        cudf_utils.HostConcatResultUtil.getColumnarBatch(hostConcatResult, dataTypes)
       }
     }
   }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/cudf_utils/HostConcatResultUtil.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/cudf_utils/HostConcatResultUtil.scala
@@ -14,9 +14,9 @@
  * limitations under the License.
  */
 
-package ai.rapids.cudf
+package com.nvidia.spark.rapids.cudf_utils
 
-import ai.rapids.cudf.JCudfSerialization.HostConcatResult
+import ai.rapids.cudf
 import com.nvidia.spark.rapids.{Arm,  GpuColumnVectorFromBuffer}
 
 import org.apache.spark.sql.types.DataType
@@ -26,11 +26,10 @@ object HostConcatResultUtil extends Arm {
   /**
    * Create a rows-only `HostConcatResult`.
    */
-  def rowsOnlyHostConcatResult(numRows: Int): HostConcatResult = {
-    new HostConcatResult(
-      new JCudfSerialization.SerializedTableHeader(
-        Array.empty, numRows, 0L),
-      HostMemoryBuffer.allocate(0, false))
+  def rowsOnlyHostConcatResult(numRows: Int): cudf.JCudfSerialization.HostConcatResult = {
+    new cudf.JCudfSerialization.HostConcatResult(
+      new cudf.JCudfSerialization.SerializedTableHeader(numRows),
+      cudf.HostMemoryBuffer.allocate(0, false))
   }
 
   /**
@@ -41,7 +40,7 @@ object HostConcatResultUtil extends Arm {
    *       callers are responsible for closing the resulting `ColumnarBatch`
    */
   def getColumnarBatch(
-      hostConcatResult: HostConcatResult,
+      hostConcatResult: cudf.JCudfSerialization.HostConcatResult,
       sparkSchema: Array[DataType]): ColumnarBatch = {
     if (hostConcatResult.getTableHeader.getNumColumns == 0) {
       // We expect the caller to have acquired the GPU unconditionally before calling

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/cudf_utils/HostConcatResultUtil.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/cudf_utils/HostConcatResultUtil.scala
@@ -16,7 +16,8 @@
 
 package com.nvidia.spark.rapids.cudf_utils
 
-import ai.rapids.cudf
+import ai.rapids.cudf.{HostMemoryBuffer, JCudfSerialization}
+import ai.rapids.cudf.JCudfSerialization.HostConcatResult
 import com.nvidia.spark.rapids.{Arm,  GpuColumnVectorFromBuffer}
 
 import org.apache.spark.sql.types.DataType
@@ -26,10 +27,10 @@ object HostConcatResultUtil extends Arm {
   /**
    * Create a rows-only `HostConcatResult`.
    */
-  def rowsOnlyHostConcatResult(numRows: Int): cudf.JCudfSerialization.HostConcatResult = {
-    new cudf.JCudfSerialization.HostConcatResult(
-      new cudf.JCudfSerialization.SerializedTableHeader(numRows),
-      cudf.HostMemoryBuffer.allocate(0, false))
+  def rowsOnlyHostConcatResult(numRows: Int): HostConcatResult = {
+    new HostConcatResult(
+      new JCudfSerialization.SerializedTableHeader(numRows),
+      HostMemoryBuffer.allocate(0, false))
   }
 
   /**
@@ -40,7 +41,7 @@ object HostConcatResultUtil extends Arm {
    *       callers are responsible for closing the resulting `ColumnarBatch`
    */
   def getColumnarBatch(
-      hostConcatResult: cudf.JCudfSerialization.HostConcatResult,
+      hostConcatResult: HostConcatResult,
       sparkSchema: Array[DataType]): ColumnarBatch = {
     if (hostConcatResult.getTableHeader.getNumColumns == 0) {
       // We expect the caller to have acquired the GPU unconditionally before calling

